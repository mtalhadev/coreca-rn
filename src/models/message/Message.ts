import { ReactionType } from '../reaction/Reaction';
import { ReactionListCLType, ReactionListType, toReactionListCLType } from '../reaction/ReactionListType';
import { ReadType } from '../read/Read';
import { ReadListCLType, ReadListType, toReadListCLType } from '../read/ReadListType';
import { RoomCLType, RoomType, toRoomCLType } from '../room/Room';
import { ThreadHeadCLType, ThreadHeadType, toThreadHeadCLType } from '../threadHead/ThreadHead';
import { toWorkerCLType, WorkerCLType, WorkerType } from '../worker/Worker';
import { CommonCLType, CommonModel, Create, ReplaceAnd, toCommonCLType, Update } from '../_others/Common';
import { ID } from '../_others/ID';
import { GetOptionObjectType, GetOptionParam } from '../_others/Option';


export type MessageEnumType = 'text' | 'picture' | 'movie' | 'note'

export type MessageModel = Partial<{
    messageId: ID;
    roomId: ID;
    threadId: ID;
    isThreadStart: boolean;
    extraRoomId: ID;
    index: number;
    workerId: ID;
    message: string;
    messageType: MessageEnumType;
    replyId: ID;
    attachmentUrl: string;
    sAttachmentUrl: string;
    xsAttachmentUrl: string;
    readCount: number;
    updateCount: number;
}> &
    CommonModel;

export const initMessage = (message: Create<MessageModel> | Update<MessageModel>): Update<MessageModel> => {
    const newMessage: Update<MessageModel> = {
        messageId: message.messageId,
        roomId: message.roomId,
        threadId: message.threadId,
        isThreadStart: message.isThreadStart,
        extraRoomId: message.extraRoomId,
        index: message.index,
        workerId: message.workerId,
        message: message.message,
        messageType: message.messageType,
        replyId: message.replyId,
        attachmentUrl: message.attachmentUrl,
        sAttachmentUrl: message.sAttachmentUrl,
        xsAttachmentUrl: message.xsAttachmentUrl,
        readCount: message.readCount,
        updateCount: message.updateCount,
    };
    return newMessage;
};

/**
 * {@link WorkerOptionInputParam - 説明}
 */
 export type MessageOptionInputParam = ReplaceAnd<
    GetOptionObjectType<MessageOptionParam>,
    {
        //
    }
>;

/**
 * {@link WorkerOptionParam - 説明}
 */
export type MessageOptionParam = {
    room?: RoomType;
    worker?: WorkerType;
    reads?: ReadListType;
    reactions?: ReactionListType;
    threadHead?: ThreadHeadType;
    reply?: MessageType;
};

export type MessageType = MessageModel & MessageOptionParam;
export type GetMessageOptionParam = GetOptionParam<MessageType, MessageOptionParam, MessageOptionInputParam>;




export type MessageCLType = ReplaceAnd<
  MessageType,
  {
    room?: RoomCLType;
    worker?: WorkerCLType;
    reads?: ReadListCLType;
    reactions?: ReactionListCLType;
    threadHead?: ThreadHeadCLType;
    reply?: MessageCLType;
  } & CommonCLType
>;

export const toMessageCLType = (data?: MessageType): MessageCLType => {
  return {
    ...data,
    ...toCommonCLType(data),
    room: data?.room ? toRoomCLType(data?.room) : undefined,
    worker: data?.worker ? toWorkerCLType(data?.worker) : undefined,
    reads: data?.reads ? toReadListCLType(data?.reads) : undefined,
    reactions: data?.reactions ? toReactionListCLType(data?.reactions) : undefined,
    threadHead: data?.threadHead ? toThreadHeadCLType(data?.threadHead) : undefined,
    reply: data?.reply ? toMessageCLType(data?.reply) : undefined,
    
  } as MessageCLType;
};