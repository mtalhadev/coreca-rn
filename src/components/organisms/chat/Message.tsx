import React from 'react'
import { Linking, Pressable, Text, View, Image, ViewStyle } from 'react-native'

import { FontStyle } from '../../../utils/Styles'
import _ from 'lodash'
import * as Clipboard from 'expo-clipboard'

import { SCREEN_WIDTH, THEME_COLORS, WINDOW_WIDTH } from '../../../utils/Constants'
import { ImageIcon } from '../ImageIcon'

import { MessageCLType } from '../../../models/message/Message'
import { useTextTranslation } from '../../../fooks/useTextTranslation'
import { Reaction } from './Reaction'
import { CustomDate, newCustomDate, timeText } from '../../../models/_others/CustomDate'
import ParsedText from 'react-native-parsed-text'
import {
    Menu,
    MenuTrigger,
    MenuOptions,
    MenuOption,
    renderers,
  } from 'react-native-popup-menu';
import { Icon } from '../../atoms/Icon'
import VectorIcon from '../../atoms/VectorIcon'
import { AppButton } from '../../atoms/AppButton'

export type MessageProps = {
    message?: MessageCLType
    myWorkerId: string
    iconSize?: number
    style?: ViewStyle
    onLongPress?: ()=> void
    onEnterThread?: ()=> void
    onImagePress?: (imageUrl: string)=> void
    onReply?: () => void
    onAddEmoji?: () => void
    onThread?: () => void
    onNotePress?: () => void
    onAddTodo?: (message?: MessageCLType) => void

}

export const Message = React.memo((props: Partial<MessageProps>) => {
    let { message, myWorkerId, style, iconSize, onLongPress, onEnterThread, onImagePress, onReply, onAddEmoji, onThread, onNotePress, onAddTodo } = props
    iconSize = iconSize ?? 25
    const _imageUri = (iconSize <= 30 ? message?.worker?.xsImageUrl : iconSize <= 50 ? message?.worker?.sImageUrl : message?.worker?.imageUrl) ?? message?.worker?.imageUrl
    const _threadHeadImageUri = (iconSize <= 30 ? message?.threadHead?.lastMessage?.worker?.xsImageUrl : iconSize <= 50 ? message?.threadHead?.lastMessage?.worker?.sImageUrl : message?.threadHead?.lastMessage?.worker?.imageUrl) ?? message?.threadHead?.lastMessage?.worker?.imageUrl

    const reactionChars: string[] = []
    const reactionCount: number[] = []

    const checkSameReaction = (reactionChar: string): [boolean, number] => {
        let hitFlag: boolean = false
        let hitIndex: number = -1

        reactionChars.forEach((char, index) => {
            if (char == reactionChar) {
                hitFlag = true
                hitIndex = index
            }
        })
        return [hitFlag, hitIndex]
    }

    const sumReaction = () => {
        message?.reactions?.items?.forEach(reaction => {
            const ret = checkSameReaction(reaction.reactionChar as string)
            if (ret[0] == false) {
                reactionChars.push(reaction.reactionChar as string)
                reactionCount.push(1)
            }
            else {
                reactionCount[ret[1]] += 1
            }
        });
    }
    sumReaction()


    /*
    if (message?.isThreadStart == true) {
        console.log(timeText(newCustomDate())  + '**' +  message.message)
    }
    */
    const _handleUrlPress = async (url: string) => {
        if(await Linking.canOpenURL(url))
            await Linking.openURL(url);
    }
    const _handleEmailPress = (email: string)  => {
        Linking.openURL(`mailto:${email}`);
    }
    const _handlePhonePress = (phone: string)  => {
        Linking.openURL(`sms:${phone}`);
    }
    const _handleMentionNamePress = (mentionName: string)  => {
    }

   
    return (
        <View
            style={[
                {
                    flexDirection: 'column',
                    borderBottomColor: THEME_COLORS.OTHERS.LIGHT_GRAY,
                    borderBottomWidth: 0.3,
                },
                style,
            ]}>
            {message?.replyId && (
                <View
                    style={{
                        backgroundColor: '#B7D1E6',
                        marginBottom: 8,
                        paddingVertical:8,
                        paddingHorizontal: 13,
                        marginLeft: -13,
                        marginRight: -10
                    }}
                >
                    <Text
                        style={{
                            color: THEME_COLORS.BLUE.MIDDLE,
                            fontSize: 12,
                            lineHeight: 14,

                        }}
                    >
                        Re : {message.reply?.worker?.name}
                    </Text>
                    <Text
                        style={{
                            color: THEME_COLORS.OTHERS.GRAY,
                            marginTop:4,
                            fontSize: 12,
                            lineHeight: 14,

                        }}
                        numberOfLines={1}
                        ellipsizeMode={'tail'}
                    >
                        {message.reply?.message}
                    </Text>

                </View>
            )}
            <Pressable
                style={{
                    flexDirection: 'row',
                    marginVertical: 8,
                }}
                onLongPress={()=> {
                    if (onLongPress) {
                        onLongPress()
                    }
                }}
            >

                <View
                    style={{
                        alignItems: 'center',
                    }}>
                    <ImageIcon imageColorHue={message?.worker?.imageColorHue} imageUri={_imageUri} type={'worker'} size={iconSize} style={{marginTop: -4}}/>
                </View>

                <View
                    style={{
                        marginLeft: 10,
                        justifyContent: 'center',
                    }}>
                    <View
                        style={{
                            flexDirection: 'column',
                            alignItems: 'flex-start',
                            paddingRight: 6,
                        }}>
                        <Text
                            numberOfLines={1}
                            ellipsizeMode={'middle'}
                            style={{
                                fontFamily: FontStyle.bold,
                                fontSize: 12,
                                lineHeight: 14,
                                width: WINDOW_WIDTH - 110,
                            }}>
                            {message?.worker?.name}
                        </Text>
                        {message?.messageType == 'text' && 
                        <ParsedText
                            childrenProps={{ ellipsizeMode: 'middle' }}
                            style={{
                                fontFamily: FontStyle.regular,
                                marginTop: 6,
                                fontSize: 12,
                                lineHeight: 16,
                                width: WINDOW_WIDTH - 110,
                            }}
                            parse={[
                                {
                                    type: 'url',
                                    style: {
                                        color: THEME_COLORS.BLUE.MIDDLE_DEEP
                                    },
                                    onPress: _handleUrlPress
                                }, {
                                    type: 'phone',
                                    style: {
                                        color: THEME_COLORS.BLUE.MIDDLE_DEEP
                                    },
                                    onPress: _handlePhonePress
                                }, {
                                    type: 'email',
                                    style: {
                                        color: THEME_COLORS.BLUE.MIDDLE_DEEP
                                    },
                                    onPress: _handleEmailPress
                                }, {
                                    pattern: /@(\w+)|@(従業員\d{5})/,
                                    style: {
                                        color: THEME_COLORS.BLUE.MIDDLE_DEEP,
                                        fontFamily: FontStyle.bold
                                    },
                                    onPress: _handleMentionNamePress
                                },
                            ]}
                            >
                                {message?.message}
                        </ParsedText>
                        }
                        {message?.messageType == 'picture' && 
                        <Pressable 
                            onPress={() => {
                                if(onImagePress)  
                                    onImagePress(message?.xsAttachmentUrl ?? 'https://no-url')
                            }} 
                        >
                            <Image
                                style={{
                                    marginTop: 6,
                                    width: SCREEN_WIDTH * 0.6,
                                    height: SCREEN_WIDTH * 0.6,
                                    resizeMode: 'contain',
                                    borderRadius: 5
                                }} 
                                source={{uri: message?.attachmentUrl ?? 'https://no-url'}}
                                testID='message-picture'
                                />
                        </Pressable>
                        }
                        {message?.messageType == 'note' && 
                        <ParsedText
                            childrenProps={{ ellipsizeMode: 'middle' }}
                            style={{
                                fontFamily: FontStyle.regular,
                                marginTop: 6,
                                fontSize: 12,
                                lineHeight: 16,
                                width: WINDOW_WIDTH - 110,
                            }}
                            parse={[
                                {
                                    type: 'url',
                                    style: {
                                        color: THEME_COLORS.BLUE.MIDDLE_DEEP
                                    },
                                    onPress: _handleUrlPress
                                }, {
                                    type: 'phone',
                                    style: {
                                        color: THEME_COLORS.BLUE.MIDDLE_DEEP
                                    },
                                    onPress: _handlePhonePress
                                }, {
                                    type: 'email',
                                    style: {
                                        color: THEME_COLORS.BLUE.MIDDLE_DEEP
                                    },
                                    onPress: _handleEmailPress
                                }, {
                                    pattern: /@(\w+)|@(従業員\d{5})/,
                                    style: {
                                        color: THEME_COLORS.BLUE.MIDDLE_DEEP,
                                        fontFamily: FontStyle.bold
                                    },
                                    onPress: _handleMentionNamePress
                                },
                            ]}
                            >
                                {message?.message}
                        </ParsedText>
                        }

                    </View>
                    {message?.messageType != 'note' &&
                    <View
                        style={{
                            flexDirection: 'row',
                            marginTop: 8,
                            width: WINDOW_WIDTH - 110,
                            flexWrap: 'wrap',
                        }}
                    >
                        {reactionChars.length > 0 && 
                            reactionChars.map((char, index) => { 
                                return <Reaction 
                                            char={char} 
                                            count={reactionCount[index]} 
                                            style = {{
                                                marginRight: 5,
                                                paddingHorizontal: 5,
                                                paddingVertical: 1,
                                                backgroundColor: THEME_COLORS.OTHERS.TABLE_AREA_PURPLE,
                                                borderRadius: 10
                                            }}
                                        />
                            })
                        }
                    </View>
                    }
                    {message?.isThreadStart && (
                        <Pressable
                            style={{
                                flexDirection: 'row',
                                alignItems: 'center',
                                marginTop: 8,
                            }}
                            onPress={() => {
                                if (onEnterThread) {
                                    onEnterThread()
                                }
                            }}
                        >
                            <ImageIcon imageColorHue={message?.threadHead?.lastMessage?.worker?.imageColorHue} imageUri={_threadHeadImageUri} type={'worker'} size={24} />
                            <Text
                                numberOfLines={1}
                                ellipsizeMode={'middle'}
                                style={{
                                    fontFamily: FontStyle.bold,
                                    color: THEME_COLORS.BLUE.MIDDLE,
                                    fontSize: 12,
                                    lineHeight: 14,
                                    marginLeft: 8,
                                    paddingTop: 4,
                                }}>
                                {message.threadHead?.messageCount}件のメッセージ
                            </Text>
                        </Pressable>
                    )}
                </View>
                <View
                    style={{
                        alignItems: 'flex-end',
                        // marginTop: 15,
                        // marginRight: 10,
                    }}
                >
                    {message?.messageType == 'note' ?
                    <AppButton
                        style={{
                            backgroundColor: "transparent",
                            paddingHorizontal: 15,
                            marginLeft: -30,
                            marginTop: 2
                        }}
                        height={30}
                        fontSize={14}
                        hasShadow={false}
                        borderColor={THEME_COLORS.OTHERS.BORDER_COLOR2}
                        borderWidth={1}
                        textColor={THEME_COLORS.OTHERS.BLACK}
                        title={'Notes'}
                        onPress={() => {
                            if(onNotePress)
                            onNotePress()
                        }}
                        />
                    :
                    <View
                        style={{
                            flexDirection: 'column',
                            alignItems: 'flex-end'
                        }}
                        
                    >
                        <Text
                            style={{
                                color: THEME_COLORS.OTHERS.GRAY,
                                fontSize: 10,
                                // backgroundColor: '#ddd',
                                marginTop: 20
                            }}
                        >
                            {timeText(message?.createdAt as CustomDate).substring(0, 5)}
                        </Text>
                        <Menu style={{
                            marginTop: 10
                        }} renderer={renderers.Popover}>
                            <MenuTrigger 
                                customStyles={{
                                    triggerWrapper: {
                                        top: -42,
                                        // right: 10
                                    },
                                }}
                                >
                                <VectorIcon.Entypo name='dots-three-horizontal' color={THEME_COLORS.OTHERS.GRAY} size={13}/>
                            </MenuTrigger>
                            <MenuOptions
                                optionsContainerStyle={{
                                    top: -48,
                                    right: 25
                                }}
                                customStyles={{
                                    optionsContainer: {
                                        justifyContent: 'center',
                                        borderRadius: 5,

                                    },
                                    optionsWrapper: {
                                        flexDirection: 'row',
                                        borderRadius: 8,
                                        backgroundColor: 'white',
                                        justifyContent: 'center',
                                        alignItems: "center",                   
                                    },
                                }}
                            >
                                <MenuOption onSelect={() => Clipboard.setStringAsync(message?.message as string)}>
                                    <Icon name='copy' fill={'#fff'} />
                                </MenuOption>
                                <MenuOption onSelect={() => onReply && onReply()}>
                                    <Icon name='reply' fill={'#fff'} />
                                </MenuOption>
                                <MenuOption onSelect={() => onThread && onThread()}>
                                    <Icon name='thread' fill={'#fff'} />
                                </MenuOption>
                                <MenuOption onSelect={() => onAddEmoji && onAddEmoji()}>
                                    <Icon name='emoji' fill={'#fff'} />
                                </MenuOption>
                                <MenuOption onSelect={() => onAddTodo && onAddTodo(message)}>
                                    <Icon name='addTodo' fill={'#fff'} />
                                </MenuOption>
                                <MenuOption >
                                    <Icon name='close' fill={THEME_COLORS.OTHERS.GRAY} width={10} style={{ marginHorizontal: 5 }}/>
                                </MenuOption>
                            </MenuOptions>
                        </Menu>
                    </View>
                    }
                    {(message?.workerId == myWorkerId && (message?.readCount ??0) > 0) &&
                    <View>
                        <Text
                            style={{
                                color: THEME_COLORS.OTHERS.GRAY,
                                fontSize: 10,
                                marginRight: 13
                            }}
                        >
                            既読{message?.readCount}
                        </Text>
                    </View>
                    }
                </View>
            </Pressable>
            
        </View>

    )
})
