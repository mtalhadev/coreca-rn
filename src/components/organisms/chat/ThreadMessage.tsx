import React, { useMemo } from 'react'
import { Pressable, Text, View, ViewStyle } from 'react-native'

import { FontStyle } from '../../../utils/Styles'
import _ from 'lodash'

import { THEME_COLORS, WINDOW_WIDTH } from '../../../utils/Constants'
import { ImageIcon } from '../ImageIcon'
import { MessageCLType, MessageType } from '../../../models/message/Message'
import { useTextTranslation } from '../../../fooks/useTextTranslation'
import { Reaction } from './Reaction'
import { Line } from '../../atoms/Line'
import { CustomDate, newCustomDate, timeText, toCustomDateFromTotalSeconds } from '../../../models/_others/CustomDate'
import { AppButton } from '../../atoms/AppButton'
import { ThreadLogType } from '../../../models/threadLog/ThreadLog'

export type ThreadMessageProps = {
    threadLog?: ThreadLogType
    iconSize?: number
    style?: ViewStyle
    onReplyPress?: ()=> void
    onReadThread?: ()=> void
}

export const ThreadMessage = React.memo((props: Partial<ThreadMessageProps>) => {
    let { threadLog, style, iconSize, onReplyPress, onReadThread } = props
    iconSize = iconSize ?? 40
    const message = threadLog?.message
    const _imageUri = (iconSize <= 30 ? message?.worker?.xsImageUrl : iconSize <= 50 ? message?.worker?.sImageUrl : message?.worker?.imageUrl) ?? message?.worker?.imageUrl
    const _threadHeadImageUri = (iconSize <= 30 ? message?.threadHead?.lastMessage?.worker?.xsImageUrl : iconSize <= 50 ? message?.threadHead?.lastMessage?.worker?.sImageUrl : message?.threadHead?.lastMessage?.worker?.imageUrl) ?? message?.threadHead?.lastMessage?.worker?.imageUrl

    const reactionChars: string[] = []
    const reactionCount: number[] = []

    const checkSameReaction = (reactionChar: string): [boolean, number] => {
        let hitFlag: boolean = false
        let hitIndex: number = -1

        reactionChars.forEach((char, index) => {
            if (char == reactionChar) {
                hitFlag = true
                hitIndex = index
            }
        })
        return [hitFlag, hitIndex]
    }

    const sumReaction = () => {
        message?.reactions?.items?.forEach(reaction => {
            const ret = checkSameReaction(reaction.reactionChar as string)
            if (ret[0] == false) {
                reactionChars.push(reaction.reactionChar as string)
                reactionCount.push(1)
            }
            else {
                reactionCount[ret[1]] += 1
            }
        });
    }
    sumReaction()


    const _getRoomSubTitle = () => {
        let rtnStr = threadLog?.room?.roomType?.toString()
        if (threadLog?.room?.roomType == 'company') {
            rtnStr = '個別 : 取引先'
        }
        if (threadLog?.room?.roomType == 'custom') {
            rtnStr = '個別 : カスタムグループ'
        }
        if (threadLog?.room?.roomType == 'onetoone') {
            rtnStr = '個別 : 個人'
        }
        if (threadLog?.room?.roomType == 'contract') {
            rtnStr = '案件/工事 : 案件'
        }
        if (threadLog?.room?.roomType == 'construction') {
            rtnStr = '案件/工事 : 工事'
        }

        return rtnStr
    }

    const createdAt = useMemo(() => message?.createdAt ? toCustomDateFromTotalSeconds(message?.createdAt) : undefined, [message?.createdAt])


    return (
        <View
            style={[
                {
                    flexDirection: 'column',
                    borderBottomColor: THEME_COLORS.OTHERS.LIGHT_GRAY,
                    borderBottomWidth: 0.3,
                },
                style,
            ]}>
            <View
                style={{}}    
            >
                <Text
                    style={{
                        fontFamily: FontStyle.bold
                    }}
                >
                    {threadLog?.room?.name}
                </Text>
                <Text
                    style={{
                        marginTop: 3,
                        fontSize: 9,


                    }}
                >
                    {_getRoomSubTitle()}
                </Text>
            </View>

            <Pressable
                style={{
                    flexDirection: 'row',
                    marginVertical: 8,
                }}
                onPress={()=> {}}
            >

                <View
                    style={{
                        alignItems: 'center',
                    }}>
                    <ImageIcon imageColorHue={message?.worker?.imageColorHue} imageUri={_imageUri} type={'worker'} size={iconSize} style={{marginTop: -4}}/>
                </View>

                <View
                    style={{
                        marginLeft: 10,
                        justifyContent: 'center',
                    }}>
                    <View
                        style={{
                            flexDirection: 'column',
                            alignItems: 'flex-start',
                            paddingRight: 6,
                        }}>
                        <Text
                            numberOfLines={1}
                            ellipsizeMode={'middle'}
                            style={{
                                fontFamily: FontStyle.bold,
                                fontSize: 12,
                                lineHeight: 14,
                                width: WINDOW_WIDTH - 110,
                            }}>
                            {message?.worker?.name}
                        </Text>
                        <Text
                            ellipsizeMode={'middle'}
                            style={{
                                fontFamily: FontStyle.regular,
                                marginTop: 6,
                                fontSize: 12,
                                lineHeight: 16,
                                width: WINDOW_WIDTH - 110,
                            }}>
                            {message?.message}
                        </Text>
                    </View>
                    <View
                        style={{
                            flexDirection: 'row',
                            marginTop: 8,
                        }}
                    >
                        {reactionChars.length > 0 && 
                            reactionChars.map((char, index) => { 
                                return <Reaction 
                                            char={char} 
                                            count={reactionCount[index]} 
                                            style={{marginRight: 10}}
                                        />
                            })
                        }
                    </View>
                </View>
                <View
                    style={{
                        alignItems: 'flex-end',
                        marginTop: 10,
                        marginRight:4,
                    }}
                >
                    <View>
                        <Text
                            style={{
                                color: THEME_COLORS.OTHERS.GRAY,
                                fontSize: 10,
                            }}
                        >
                            {timeText(createdAt ?? newCustomDate()).substring(0, 5)}
                        </Text>
                    </View>
                    <View>
                        <Text
                            style={{
                                color: THEME_COLORS.OTHERS.GRAY,
                                fontSize: 10,
                            }}
                        >
                            既読{message?.readCount}
                        </Text>
                    </View>
                </View>
            </Pressable>
            <View
                style={{
                    marginTop: 0,
                    flexDirection: 'column',
                }}
            >
                <Pressable
                    style={{
                        marginBottom: 7,
                    }}
                    onPress={()=> {
                        if (onReadThread) {
                            onReadThread()
                        }
                    }}
                >
                    <Text
                        numberOfLines={1}
                        ellipsizeMode={'middle'}
                        style={{
                            fontFamily: FontStyle.bold,
                            color: THEME_COLORS.BLUE.MIDDLE,
                            fontSize: 12,
                            lineHeight: 14,
                            marginLeft: 0,
                            paddingTop: 4,
                        }}>
                        その他のメッセージを読む
                    </Text>
                </Pressable>

                <AppButton
                    isGray={true}
                    height={30}
                    fontSize={13}
                    title={'返信する'}
                    style={{
                        flex: 1,
                        marginLeft: 0,
                        marginBottom: 20,
                        marginTop: 8,
                        width: 90,
                    }}
                    onPress={() => {
                        if (onReplyPress) {
                            onReplyPress()
                        }
                    }}
                />

            </View>
            
        </View>

    )
})
