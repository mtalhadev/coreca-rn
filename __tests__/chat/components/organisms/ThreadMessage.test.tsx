import { render, screen } from '@testing-library/react-native';
import React from 'react';
import { Message } from '../../../../src/components/organisms/chat/Message';
import { CustomDate } from '../../../../src/models/_others/CustomDate';
import { MenuProvider } from 'react-native-popup-menu';
import { GreenColor } from '../../../../src/utils/Styles';
import { ThreadMessage } from '../../../../src/components/organisms/chat/ThreadMessage';

describe('ThreadMessage test', () => {

    beforeEach(() => {
      jest.resetAllMocks()
    })
  
    it('renders Message type - text', () => {
      const date: CustomDate = new Date(2000,0,1,11,30,0).toCustomDate()

        render(
            <ThreadMessage 
              threadLog={{
                message: {
                  messageId: 'message-id',
                  roomId: 'room-id',
                  threadId: 'thread-id',
                  message: "@TestUser It's a test",
                  messageType: 'text',
                  createdAt: date.totalSeconds,
                  room: {
                    name: 'Room-XYZ',
                    roomType: "project"
                  },
                  worker: {
                    name: 'ABC',
                    xsImageUrl: 'https://profile-image-xs',
                    sImageUrl: 'https://profile-image-s',
                    imageUrl: 'https://profile-image',
                    imageColorHue: 40
                  }
                }
              }}
            />
        );

        expect(screen.getByText('ABC')).toBeVisible();
        expect(screen.getByText('その他のメッセージを読む')).toBeVisible();
        expect(screen.getByText("@TestUser It's a test")).toBeVisible();
        expect(screen.getByText("11:30")).toBeVisible(); 
        expect(screen.getByText("返信する")).toBeVisible(); 
        expect(screen.getByText("既読")).toBeVisible(); 
        const image = screen.getByTestId("ImageIcon-Image")
        expect(image.props.source.uri).toBe("https://profile-image-s");
        expect(image.props.style.width).toBe(40);
        expect(image.props.style.height).toBe(40);
        const imageBorder = screen.getByTestId("ImageIcon-Border")
        expect(imageBorder.props.style.borderColor).toBe(GreenColor.subColor);       
        expect(screen).toMatchSnapshot();    
    }) 

})  