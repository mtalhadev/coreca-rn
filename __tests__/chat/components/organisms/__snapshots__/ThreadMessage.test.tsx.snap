// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThreadMessage test renders Message type - text 1`] = `
<View
  style={
    Array [
      Object {
        "borderBottomColor": "#aaa",
        "borderBottomWidth": 0.3,
        "flexDirection": "column",
      },
      undefined,
    ]
  }
>
  <View
    style={Object {}}
  >
    <Text
      style={
        Object {
          "fontFamily": "NotoSansJP_700Bold",
        }
      }
    />
    <Text
      style={
        Object {
          "fontSize": 9,
          "marginTop": 3,
        }
      }
    />
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "flexDirection": "row",
        "marginVertical": 8,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#888",
            "borderBottomLeftRadius": 40,
            "borderBottomRightRadius": 40,
            "borderRadius": 40,
            "borderTopLeftRadius": 40,
            "borderTopRightRadius": 40,
            "height": 40,
            "justifyContent": "center",
            "marginTop": -4,
            "width": 40,
          }
        }
      >
        <RNCMaskedView>
          <View
            pointerEvents="none"
            style={
              Object {
                "bottom": 0,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              }
            }
          >
            <View
              style={
                Object {
                  "backgroundColor": "#000",
                  "borderRadius": 40,
                  "height": 40,
                  "width": 40,
                }
              }
            />
          </View>
          <Image
            resizeMode="cover"
            source={
              Object {
                "uri": "https://profile-image-s",
              }
            }
            style={
              Object {
                "height": 40,
                "width": 40,
              }
            }
            testID="ImageIcon-Image"
          />
        </RNCMaskedView>
        <View
          style={
            Object {
              "borderColor": "#C7D601",
              "borderRadius": 40,
              "borderWidth": 2,
              "height": 40,
              "position": "absolute",
              "width": 40,
            }
          }
          testID="ImageIcon-Border"
        />
      </View>
    </View>
    <View
      style={
        Object {
          "justifyContent": "center",
          "marginLeft": 10,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "flex-start",
            "flexDirection": "column",
            "paddingRight": 6,
          }
        }
      >
        <Text
          ellipsizeMode="middle"
          numberOfLines={1}
          style={
            Object {
              "fontFamily": "NotoSansJP_700Bold",
              "fontSize": 12,
              "lineHeight": 14,
              "width": 640,
            }
          }
        >
          ABC
        </Text>
        <Text
          ellipsizeMode="middle"
          style={
            Object {
              "fontFamily": "NotoSansJP_400Regular",
              "fontSize": 12,
              "lineHeight": 16,
              "marginTop": 6,
              "width": 640,
            }
          }
        >
          @TestUser It's a test
        </Text>
      </View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "marginTop": 8,
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "alignItems": "flex-end",
          "marginRight": 4,
          "marginTop": 10,
        }
      }
    >
      <View>
        <Text
          style={
            Object {
              "color": "#737680",
              "fontSize": 10,
            }
          }
        >
          11:30
        </Text>
      </View>
      <View>
        <Text
          style={
            Object {
              "color": "#737680",
              "fontSize": 10,
            }
          }
        >
          既読
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "flexDirection": "column",
        "marginTop": 0,
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "marginBottom": 7,
        }
      }
    >
      <Text
        ellipsizeMode="middle"
        numberOfLines={1}
        style={
          Object {
            "color": "#0082C7",
            "fontFamily": "NotoSansJP_700Bold",
            "fontSize": 12,
            "lineHeight": 14,
            "marginLeft": 0,
            "paddingTop": 4,
          }
        }
      >
        その他のメッセージを読む
      </Text>
    </View>
    <View
      collapsable={false}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "backgroundColor": "#F7F7F8",
          "borderColor": "#aaa",
          "borderRadius": 30,
          "borderWidth": 1,
          "elevation": 0,
          "flex": 1,
          "height": 30,
          "justifyContent": "center",
          "marginBottom": 20,
          "marginLeft": 0,
          "marginTop": 8,
          "opacity": 1,
          "shadowColor": "#000",
          "shadowOffset": Object {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0,
          "shadowRadius": 8,
          "width": 90,
        }
      }
    >
      <View
        accessibilityState={
          Object {
            "disabled": false,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": undefined,
              "paddingHorizontal": undefined,
            }
          }
        >
          <View
            style={
              Object {
                "alignContent": undefined,
                "flexDirection": undefined,
                "justifyContent": undefined,
                "paddingTop": 0,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#343538",
                  "fontFamily": "NotoSansJP_400Regular",
                  "fontSize": 13,
                  "lineHeight": 30,
                  "paddingTop": 0,
                  "textAlign": undefined,
                }
              }
            >
              返信する
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
